thewalrus/fock_gradients.py

```diff
@@ -355,3 +357,89 @@ def grad_beamsplitter(T, theta, phi):  # pragma: no cover
                     )
 
     return grad_theta, grad_phi
+
+
+@jit(nopython=True)
+def mzgate(theta, phi, cutoff, dtype=np.complex128):  # pragma: no cover
+    r"""Calculates the Fock representation of the mzgate.
+
+    Args:
+        theta (float): transmissivity angle of the mzgate. The transmissivity is :math:`t=\cos(\theta)`
+        phi (float): reflection phase of the mzgate
+        cutoff (int): Fock ladder cutoff
+        dtype (data type): Specifies the data type used for the calculation
+
+    Returns:
+        array[float]: The Fock representation of the gate
+    """
+    sqrt = np.sqrt(np.arange(cutoff, dtype=dtype))
+    v = np.exp(1j * theta)
+    u = np.exp(1j * phi)
+    R = 0.5*np.array([[0, 0, u*(v-1), 1j*(1+v)], [0, 0, 1j*u*(1+v), 1-v], [u*(v-1), 1j*u*(1+v), 0, 0], [1j*(1+v), 1-v, 0, 0],])
+
+    Z = np.zeros((cutoff, cutoff, cutoff, cutoff), dtype=dtype)
+    Z[0, 0, 0, 0] = 1.0
+
+    # rank 3
+    for m in range(cutoff):
+        for n in range(cutoff - m):
+            p = m + n
+            if 0 < p < cutoff:
```

Comment:
Maybe it's not possible, but I'm curious; in this loop (and others below), is it possible to use the condition to reduce the number of for loops? If so, does that make the execution faster?

Reply:
I think we tried this in the past, and trying to roll into the loop the selection rule did not help much wit speed.